### PR TITLE
[FEATURE] add support for Azure authentication methods

### LINKED ISSUE #3089: Add service principal authentication support for `TupleAzureBlobStoreBackend`
Currently, `TupleAzureBlobStoreBackend` only supports blob connection string (`AZURE_STORAGE_CONNECTION_STRING`). However, in many companies, the preferred way to authenticate is through service accounts ("service principals" as Msoft calls them).

Could possibly be done using [adlfs](https://github.com/dask/adlfs):
```python
from adlfs import AzureBlobFileSystem

fs = AzureBlobFileSystem(
    account_name=storage_account_name,
    tenant_id=tenant_id,
    client_id=client_id,
    client_secret=client_secret,
)
fs.cp(from_path, to_path)
fs.exists(path)
fs.download(rpath, lpath)
...
```

### LINKED ISSUE #4920: Azure Blob Storage: Shared access signature (SAS) authorisation
Hi all,

I am intending to use Great Expectations on data that is stored in Azure Blob Storage (ASB) container using SAS token instead of access key. 

I have followed these instructions: https://docs.greatexpectations.io/docs/guides/connecting_to_your_data/cloud/azure/spark

What I have established is that when account_url/account_key pair is specified in data_connectors part of the *.yaml configuration file (see below), verification of the new datasource as described in [step 5 of the tutorial](https://docs.greatexpectations.io/docs/guides/connecting_to_your_data/cloud/azure/spark#5-test-your-new-datasource) succeeds. 

However, when account_url/SAS token pair is provided, I get:

 `java.lang.IllegalArgumentException: The String is not a valid Base64-encoded string.` error.

```
example_yaml = rf"""
name: my_azure_datasource
class_name: Datasource
execution_engine:
    class_name: SparkDFExecutionEngine
    azure_options:
        account_url: <mystorageaccount>.blob.core.windows.net
        credential: <SAS_TOKEN_OR_ACCOUNT_KEY>
data_connectors:
    default_runtime_data_connector_name:
        class_name: RuntimeDataConnector
        batch_identifiers:
            - default_identifier_name
    default_inferred_data_connector_name:
        class_name: InferredAssetAzureDataConnector
        azure_options:
            account_url: <mystorageaccount>.blob.core.windows.net
            credential: <SAS_TOKEN_OR_ACCOUNT_KEY>
        container: my-container
        name_starts_with: my_data
        default_regex:
            pattern: (.*)\.parquet
            group_names:
                - data_asset_name
"""
```

Furthermore, continuing with account_url/account_key pair, I tried to create an expectation suite using following commands:

```
**~$ great_expectations suite new**
Using v3 (Batch Request) API
Using Spark's default log4j profile: org/apache/spark/log4j-defaults.properties
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
22/04/21 16:36:38 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
22/04/21 16:36:39 WARN Utils: Service 'SparkUI' could not bind on port 4040. Attempting port 4041.
22/04/21 16:36:39 WARN Utils: Service 'SparkUI' could not bind on port 4041. Attempting port 4042.

**How would you like to create your Expectation Suite?**
    1. Manually, without interacting with a sample batch of data (default)
    2. Interactively, with a sample batch of data
    3. Automatically, using a profiler
: 3

**A batch of data is required to edit the suite - let's help you to specify it.**

**Which data asset (accessible by data connector "default_inferred_data_connector_name") would you like to use?**
    1. somefile

Type [n] to see the next page or [p] for the previous. When you're ready to select an asset, enter the index.
: 1

Name the new Expectation Suite [somefile]: conf_test

Great Expectations will create a notebook, containing code cells that select from available columns in your dataset and
generate expectations about them to demonstrate some examples of assertions you can make about your data.

When you run this notebook, Great Expectations will store these expectations in a new Expectation Suite "exp_test" here:

  file:///home/<USER_NAME>/great_expectations/expectations/exp_test.json

Would you like to proceed? [Y/n]: Y
22/04/21 16:39:22 WARN FileStreamSink: Assume no metadata directory. Error while looking for metadata directory in the path: wasbs://my-file-system@<mystorageaccount>.blob.core.windows.net/somefile.csv.
java.lang.RuntimeException: java.lang.ClassNotFoundException: Class org.apache.hadoop.fs.azure.NativeAzureFileSystem$Secure not found
        at org.apache.hadoop.conf.Configuration.getClass(Configuration.java:2667)
        at org.apache.hadoop.fs.FileSystem.getFileSystemClass(FileSystem.java:3431)
        at org.apache.hadoop.fs.FileSystem.createFileSystem(FileSystem.java:3466)
        at org.apache.hadoop.fs.FileSystem.access$300(FileSystem.java:174)
        at org.apache.hadoop.fs.FileSystem$Cache.getInternal(FileSystem.java:3574)
        at org.apache.hadoop.fs.FileSystem$Cache.get(FileSystem.java:3521)
        at org.apache.hadoop.fs.FileSystem.get(FileSystem.java:540)
        at org.apache.hadoop.fs.Path.getFileSystem(Path.java:365)
        at org.apache.spark.sql.execution.streaming.FileStreamSink$.hasMetadata(FileStreamSink.scala:53)
        at org.apache.spark.sql.execution.datasources.DataSource.resolveRelation(DataSource.scala:370)
        at org.apache.spark.sql.DataFrameReader.loadV1Source(DataFrameReader.scala:274)
        at org.apache.spark.sql.DataFrameReader.$anonfun$load$3(DataFrameReader.scala:245)
        at scala.Option.getOrElse(Option.scala:189)
        at org.apache.spark.sql.DataFrameReader.load(DataFrameReader.scala:245)
        at org.apache.spark.sql.DataFrameReader.csv(DataFrameReader.scala:571)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
        at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
        at py4j.Gateway.invoke(Gateway.java:282)
        at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
        at py4j.commands.CallCommand.execute(CallCommand.java:79)
        at py4j.ClientServerConnection.waitForCommands(ClientServerConnection.java:182)
        at py4j.ClientServerConnection.run(ClientServerConnection.java:106)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.ClassNotFoundException: Class org.apache.hadoop.fs.azure.NativeAzureFileSystem$Secure not found
        at org.apache.hadoop.conf.Configuration.getClassByName(Configuration.java:2571)
        at org.apache.hadoop.conf.Configuration.getClass(Configuration.java:2665)
        ... 26 more
Traceback (most recent call last):
  File "/opt/conda/bin/great_expectations", line 10, in <module>
    sys.exit(main())
  File "/opt/conda/lib/python3.7/site-packages/great_expectations/cli/cli.py", line 184, in main
    cli()
  File "/opt/conda/lib/python3.7/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/opt/conda/lib/python3.7/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/opt/conda/lib/python3.7/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/opt/conda/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/opt/conda/lib/python3.7/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/click/decorators.py", line 21, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/great_expectations/cli/suite.py", line 143, in suite_new
    batch_request=batch_request,
  File "/opt/conda/lib/python3.7/site-packages/great_expectations/cli/suite.py", line 318, in _suite_new_workflow
    raise e
  File "/opt/conda/lib/python3.7/site-packages/great_expectations/cli/suite.py", line 294, in _suite_new_workflow
    assume_yes=False,
  File "/opt/conda/lib/python3.7/site-packages/great_expectations/cli/suite.py", line 844, in _suite_edit_workflow
    raise e
  File "/opt/conda/lib/python3.7/site-packages/great_expectations/cli/suite.py", line 781, in _suite_edit_workflow
    batch_request=batch_request,
  File "/opt/conda/lib/python3.7/site-packages/great_expectations/render/renderer/v3/suite_profile_notebook_renderer.py", line 40, in __init__
    expectation_suite_name=expectation_suite_name,
  File "/opt/conda/lib/python3.7/site-packages/great_expectations/data_context/data_context.py", line 1862, in get_validator
    **kwargs,
  File "/opt/conda/lib/python3.7/site-packages/great_expectations/core/usage_statistics/usage_statistics.py", line 287, in usage_statistics_wrapped_method
    result = func(*args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/great_expectations/data_context/data_context.py", line 1749, in get_batch_list
    return datasource.get_batch_list_from_batch_request(batch_request=batch_request)
  File "/opt/conda/lib/python3.7/site-packages/great_expectations/datasource/new_datasource.py", line 195, in get_batch_list_from_batch_request
    batch_definition=batch_definition
  File "/opt/conda/lib/python3.7/site-packages/great_expectations/datasource/data_connector/data_connector.py", line 111, in get_batch_data_and_metadata
    batch_spec=batch_spec
  File "/opt/conda/lib/python3.7/site-packages/great_expectations/execution_engine/sparkdf_execution_engine.py", line 264, in get_batch_data_and_markers
    batch_data = reader_fn(path)
  File "/opt/conda/lib/python3.7/site-packages/pyspark/sql/readwriter.py", line 410, in csv
    return self._df(self._jreader.csv(self._spark._sc._jvm.PythonUtils.toSeq(path)))
  File "/opt/conda/lib/python3.7/site-packages/py4j/java_gateway.py", line 1322, in __call__
    answer, self.gateway_client, self.target_id, self.name)
  File "/opt/conda/lib/python3.7/site-packages/pyspark/sql/utils.py", line 111, in deco
    return f(*a, **kw)
  File "/opt/conda/lib/python3.7/site-packages/py4j/protocol.py", line 328, in get_return_value
    format(target_id, ".", name), value)
py4j.protocol.Py4JJavaError: An error occurred while calling o113.csv.
: java.lang.RuntimeException: java.lang.ClassNotFoundException: Class org.apache.hadoop.fs.azure.NativeAzureFileSystem$Secure not found
        at org.apache.hadoop.conf.Configuration.getClass(Configuration.java:2667)
        at org.apache.hadoop.fs.FileSystem.getFileSystemClass(FileSystem.java:3431)
        at org.apache.hadoop.fs.FileSystem.createFileSystem(FileSystem.java:3466)
        at org.apache.hadoop.fs.FileSystem.access$300(FileSystem.java:174)
        at org.apache.hadoop.fs.FileSystem$Cache.getInternal(FileSystem.java:3574)
        at org.apache.hadoop.fs.FileSystem$Cache.get(FileSystem.java:3521)
        at org.apache.hadoop.fs.FileSystem.get(FileSystem.java:540)
        at org.apache.hadoop.fs.Path.getFileSystem(Path.java:365)
        at org.apache.spark.sql.execution.datasources.DataSource$.$anonfun$checkAndGlobPathIfNecessary$1(DataSource.scala:747)
        at scala.collection.immutable.List.map(List.scala:293)
        at org.apache.spark.sql.execution.datasources.DataSource$.checkAndGlobPathIfNecessary(DataSource.scala:745)
        at org.apache.spark.sql.execution.datasources.DataSource.checkAndGlobPathIfNecessary(DataSource.scala:577)
        at org.apache.spark.sql.execution.datasources.DataSource.resolveRelation(DataSource.scala:408)
        at org.apache.spark.sql.DataFrameReader.loadV1Source(DataFrameReader.scala:274)
        at org.apache.spark.sql.DataFrameReader.$anonfun$load$3(DataFrameReader.scala:245)
        at scala.Option.getOrElse(Option.scala:189)
        at org.apache.spark.sql.DataFrameReader.load(DataFrameReader.scala:245)
        at org.apache.spark.sql.DataFrameReader.csv(DataFrameReader.scala:571)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
        at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
        at py4j.Gateway.invoke(Gateway.java:282)
        at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
        at py4j.commands.CallCommand.execute(CallCommand.java:79)
        at py4j.ClientServerConnection.waitForCommands(ClientServerConnection.java:182)
        at py4j.ClientServerConnection.run(ClientServerConnection.java:106)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.ClassNotFoundException: Class org.apache.hadoop.fs.azure.NativeAzureFileSystem$Secure not found
        at org.apache.hadoop.conf.Configuration.getClassByName(Configuration.java:2571)
        at org.apache.hadoop.conf.Configuration.getClass(Configuration.java:2665)
        ... 29 more
```

Hence, I have two questions:

1. Is it possible to use SAS tokens in order to access data stored in ABS container or only account_url/account_key pair is allowed?

2. Given successful configuration of the data connector using account_url/account_key pair, what might cause problems during suite creation?

Kind regards

### PR DESCRIPTION
**Pull request by sdebruyn**
This adds support for all common Azure authentication methods for metadata and data docs backends.

Changes proposed in this pull request:
- Fixes behaviour of `AZURE_STORAGE_CONNECTION_STRING` env var so that it matches behaviour described in the docs
- Add support for all common Azure authentication methods
- Document authentication options
- Apparently the unit tests for this backend were always skipped, so fixed them and added new ones to cover the above

Closes #3089

### Definition of Done

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

### CODE REVIEW

**Comment by sdebruyn** [comment]
Apparently the current tests covering this class never run because of a coding error. I am fixing this right now. Please wait with merging.

File: requirements-dev-azure.txt
Comment (AFineDayFor) [inline requirements-dev-azure.txt:1]: NIT: This needs to be looked into or addressed but as it's in dev it's likely a lgtm.

File: great_expectations/data_context/store/tuple_store_backend.py
Comment (AFineDayFor) [inline great_expectations/data_context/store/tuple_store_backend.py:974]: raising with team.

File: great_expectations/data_context/store/tuple_store_backend.py
Comment (AFineDayFor) [inline great_expectations/data_context/store/tuple_store_backend.py:1000]: NIT: deviates from scheme upheld in other locations.

**Review by AFineDayFor — COMMENTED** [review]
Howdy @sdebruyn, thank you very much for taking time to address this item 🎉 

We'll go over the addition of the `AZURE_STORAGE_ACCOUNT_URL` var in respect to `AZURE_STORAGE_CONNECTION_STRING`with the team.

File: great_expectations/data_context/store/tuple_store_backend.py
Comment (sdebruyn) [reply great_expectations/data_context/store/tuple_store_backend.py:1000]: What would you like me to name it here? Since it's a property now, the get should not be here

File: requirements-dev-azure.txt
Comment (sdebruyn) [reply requirements-dev-azure.txt:1]: Ah I should also have updated this in the actual dependencies. You'd need this if you want to use this backend. Most Azure users have it installed anyway and I think it's even a dependency of the storage SDK.

Comment (sdebruyn) [reply requirements-dev-azure.txt:1]: Unclear where I have to define this

**Comment by sdebruyn** [comment]
> Howdy @sdebruyn, thank you very much for taking time to address this item 🎉
> 
> We'll go over the addition of the `AZURE_STORAGE_ACCOUNT_URL` var in respect to `AZURE_STORAGE_CONNECTION_STRING`with the team.

Regarding these 2 environment variables:
* Microsoft uses the same environment variable in their docs: https://docs.microsoft.com/en-us/azure/storage/blobs/storage-quickstart-blobs-python?tabs=environment-variable-windows#configure-your-storage-connection-string and it is also used in adlfs: https://github.com/fsspec/adlfs/blob/37324d0e952fd2f5f7b382f453fbcac99e2fb130/adlfs/spec.py#L426
* The account url env var is also used in this sample by Microsoft: https://github.com/Azure/azure-sdk-for-python/blob/393d86c6f4f0b7ad5ff47496328414d92c448c0e/sdk/storage/azure-storage-file-share/samples/file_samples_authentication_async.py#L37

**Comment by psxmc6** [comment]
Hi, 

When we could expect this feature to be released?

Kind regards

**Comment by kyleaton** [comment]
Hi @psxmc6,

We have this item is high in our queue for items to address with our core engineering team but some key decision-makers are unavailable this week so we should have an update by this time next week at the latest.

**Comment by sdebruyn** [comment]
@kyleaton Are the decision makers back? I don't think this is a controversial PR :)

**Comment by psxmc6** [comment]
Looking forward to it as well!

**Review by AFineDayFor — DISMISSED** [review]

File: docs/guides/setup/configuring_data_docs/how_to_host_and_share_data_docs_on_azure_blob_storage.md
Comment (Shinnnyshinshin) [inline docs/guides/setup/configuring_data_docs/how_to_host_and_share_data_docs_on_azure_blob_storage.md:67]: 🙇🏼

File: docs/guides/setup/configuring_data_docs/how_to_host_and_share_data_docs_on_azure_blob_storage.md
Comment (Shinnnyshinshin) [inline docs/guides/setup/configuring_data_docs/how_to_host_and_share_data_docs_on_azure_blob_storage.md:91]: Thank you very much for this additional bit of detail 🙇🏼

File: great_expectations/data_context/store/tuple_store_backend.py
Comment (Shinnnyshinshin) [reply great_expectations/data_context/store/tuple_store_backend.py:974]: Looks good to me, although I'd request typehinting `Optional[str]` to these parameters to make it more explicit.

File: requirements-dev-azure.txt
Comment (Shinnnyshinshin) [reply requirements-dev-azure.txt:1]: Looks good to me. @sdebruyn, thank you for the update.

**Review by Shinnnyshinshin — APPROVED** [review]
Really appreciate the work you put into this @sdebruyn. LGTM

File: great_expectations/data_context/store/tuple_store_backend.py
Comment (Shinnnyshinshin) [reply great_expectations/data_context/store/tuple_store_backend.py:974]: On second thought, this is probably on us :) Don't worry about this for your PR @sdebruyn

**Comment by sdebruyn** [comment]
Rebased on master

**Comment by sdebruyn** [comment]
@Shinnnyshinshin @AFineDayFor What do I need to do to get this merged?

**Comment by psxmc6** [comment]
Is there any update on this pull request?

**Comment by sdebruyn** [comment]
I am waiting for one of the maintainers to merge this. I am available to make any changes if needed

**Comment by Shinnnyshinshin** [comment]
@sdebruyn looks like this PR merged successfully :) thank you for contribution, and your patience

**Comment by psxmc6** [comment]
Hi, I am still unsure how to correctly define SAS token in order to access data source. The following configuration throws an error:

image

Error:

image

Appreciate your help!

**Comment by sdebruyn** [comment]
Hi @psxmc6 you would have to put a connection string to use SAS authentication. The property is called `connection_string` and the value is in the form of `BlobEndpoint=https://youraccountname.blob.core.windows.net/;SharedAccessSignature=sv=2021-06-08...`

**Pull request merged**
